build: create JS bundles for each component

jelbourn opened the pull request:
Before this change, all of `@angular/material` was bundled into a
single, flat JS file.

After this change, each top-level component (i.e., each directory under
`src/lib`) will be have its own JS file (and d.ts and metadata.json).

The benefit of this change is that current versions of webpack will be
able to do a much better job of only including the code an app is using.

It also allows (but does not require) people to explicitly import from specific components,
e.g.
```ts
import {MdButtonModule} from '@angular/material/button';
```

For development _within Angular Material itself_ we have to use these
deeper imports because the root `@angular/material` doesn't exist until
the actual package is created.

Additionally, this PR adds a `tsconfig.json` for each dir under `src/`
with the correct path mappings for that directory, so that IDEs will
correctly generate imports.

Closure Compiler CI check is also disabled by this PR; it cannot work
with the new package structure until a release with
google/closure-compiler#2600 is made available.

Fixes #4137

DevVersion commented on src/lib/tsconfig-build.json:
```diff
@@ -5,7 +5,7 @@
     "stripInternal": false,
     "experimentalDecorators": true,
-    "noUnusedParameters": true,
+    "noUnusedParameters": false,
```
Why are we no longer having this check? Maybe add a comment if necessary

jelbourn replied:
Was just trying something during development; reverted

DevVersion commented on tools/package-tools/entry-point-package-json.ts:
```diff
@@ -8,8 +8,8 @@ export function createEntryPointPackageJson(destDir: string, packageName: string
-    module: `../@angular/${packageName}/${entryPointName}.es5.js`,
-    es2015: `../@angular/${packageName}/${entryPointName}.js`,
+    module: `../esm5/${packageName}/${entryPointName}.es5.js`,
+    es2015: `../esm2015/${packageName}/${entryPointName}.js`,
```
Out of curiosity: How does this work with Closure Compiler now? Is there an alternative to the `js_module_roots` option?

jelbourn replied:
Short answer: it doesn't.

I updated the script and left a comment explaining the current situation w/ closure.

DevVersion commented on src/lib/tsconfig-umd.json:
```diff
@@ -0,0 +1,39 @@
+    "flatModuleId": "@angular/material",
+    "skipTemplateCodegen": true
+  }
+}
```
Is that file used anywhere? We now have three `tsconfig-XX` files in the `lib/` package.

jelbourn replied:
I was just trying something out with this, meant to delete it.

jelbourn commented:
* Rebased such that it includes `material-moment-adapter/`
* Added a `tsconfig.json` per `src/` directory in order to make IDEs auto-add imports to the right path.
* Disabled closure compiler CI check temporarily (need a new release)
* Refactored rollup bundling code into a class, `PackageBundler` so that it can maintain the context of the `BuildPackage`
* Fix prerender

jelbourn commented:
@DevVersion after this PR we're definitely going to have to parallelize the typescript builds and bundling. With this change it takes over a minute. To paralleize the tasks, we'll need to manually spawn a child process for each, since nodejs is single-threaded. We also probably want to tweak the watch tasks to only re-compile the sub-packages that changed. Let me know if you want to work on all that.



All of this will be obsolete when we switch to bazel, but we're at least a month away from making that switch.

jelbourn commented:
@DevVersion the tsconfig in the root would cause IDEs to autogenerate the wrong imports sometimes.

jelbourn commented:
Caretaker note: both the copy config and the build rules need to updated alongside this change

jelbourn commented:
Another note: if this gets merged alongside another PR, it will potentially need a build fix (for anything importing from, e.g., `../core`)
